pkg/controllers/provisioningv2/rke2/machinedrain/machinedrain.go

```diff
@@ -244,3 +240,17 @@ func (h *handler) performDrain(machine *capi.Machine, drainOpts *rkev1.DrainOpti
 
 	return drain.RunNodeDrain(helper, node.Name)
 }
+
+func (h *handler) updateSecretAnnotation(secret *corev1.Secret, annotation, value string) (*corev1.Secret, error) {
+	var err error
+	return secret, retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
I wonder if using a `retry` here can run us into weird edge cases where the drain/uncordon annotations are being manipulated in rapid succession.

If anything, I would almost want to do a `retry.RetryOnConflict` where we validate that the annotations we care about (the drain/uncordon ones) are in the "old" set state before we update them -- we should easily be able to pass in the current values/keys we care about to compare against.

Reply:
The reason I had to add this retry is because the `system-agent` is also updating the plan secret every second while the upgrade/drain is happening. I am unsure if removing this will allow the updating of the secret to be successful.

Reply:
I'm not saying we should remove the retry, I'm saying we should validate the before settings of the annotations being what we expect before we manipulate the latest secret we retrieve.

Reply:
Ah. I misunderstood. Yes, I will add this.

Reply:
Sorry if I was unclear, this is actually not exactly what I meant. 

In this specific controller, we are looking for and manipulate a couple of annotations, i.e. `rke2.DrainDoneAnnotation`, `rke2.PreDrainAnnotation`, `rke2.PostDrainAnnotation`, `rke2.DrainErrorAnnotation`, `rke2.UncordonAnnotation`, etc. 

What I worry about is that this controller is taking a specific version of the secret, it is then processing the "input" annotations i.e. `rke2.DrainAnnotation` and then setting a bunch of other annotations based on that input. 

The issue with the `retry` here is that, let's say we have a situation where the `rke2.DrainAnnotation` is set, then immediately updated, the "input" secret is no longer the same as what we are currently processing and the output i.e. result of the event is going to have whatever specific annotations we're trying to set on the secret. Because the secret will have been updated while the secret is being processed through this controller, we will get a conflict and will be unable to set the annotation, but will continuously retry to set that annotation reflecting the "old" condition even though that has been updated at this point.

What I am thinking is before allowing the retry to continue, we should validate that our input, i.e. `rke2.DrainAnnotation`, `hook.Annotaiton`, etc. are all what we expect them to be every time we retrieve the secret from the client before setting our "result" annotations. This is to help prevent unnecessary thrashing of the controller, and keeps us closer to the conventional model of "re-enqueue if conflict"